// @pnpm/plugin-commands-rebuild/lib/implementation/extendRebuildOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const normalize_registries_1 = require("@pnpm/normalize-registries");
const loadJsonFile = require("load-json-file");
const path = require("path");
const defaults = async (opts) => {
    const packageManager = opts.packageManager ||
        await loadJsonFile(path.join(__dirname, '../../package.json'));
    const dir = opts.dir || process.cwd();
    const lockfileDir = opts.lockfileDir || dir;
    return {
        childConcurrency: 5,
        development: true,
        dir,
        force: false,
        forceSharedLockfile: false,
        lockfileDir,
        optional: true,
        packageManager,
        pending: false,
        production: true,
        rawConfig: {},
        registries: normalize_registries_1.DEFAULT_REGISTRIES,
        shamefullyHoist: false,
        sideEffectsCacheRead: false,
        storeDir: opts.storeDir,
        unsafePerm: process.platform === 'win32' ||
            process.platform === 'cygwin' ||
            !(process.getuid && process.setuid &&
                process.getgid && process.setgid) ||
            process.getuid() !== 0,
        useLockfile: true,
        userAgent: `${packageManager.name}/${packageManager.version} npm/? node/${process.version} ${process.platform} ${process.arch}`,
    };
};
exports.default = async (opts) => {
    if (opts) {
        for (const key in opts) {
            if (opts[key] === undefined) {
                delete opts[key];
            }
        }
    }
    const defaultOpts = await defaults(opts);
    const extendedOpts = { ...defaultOpts, ...opts, storeDir: defaultOpts.storeDir };
    extendedOpts.registries = normalize_registries_1.default(extendedOpts.registries);
    return extendedOpts;
};
