// pnpm/lib/cmd/installTest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cli_utils_1 = require("@pnpm/cli-utils");
const plugin_commands_installation_1 = require("@pnpm/plugin-commands-installation");
const plugin_commands_script_runners_1 = require("@pnpm/plugin-commands-script-runners");
const renderHelp = require("render-help");
exports.cliOptionsTypes = plugin_commands_installation_1.install.cliOptionsTypes;
exports.rcOptionsTypes = plugin_commands_installation_1.install.rcOptionsTypes;
exports.commandNames = ['install-test', 'it'];
function help() {
    return renderHelp({
        aliases: ['it'],
        description: 'Runs a \`pnpm install\` followed immediately by a \`pnpm test\`. It takes exactly the same arguments as \`pnpm install\`.',
        url: cli_utils_1.docsUrl('install-test'),
        usages: ['pnpm install-test'],
    });
}
exports.help = help;
async function handler(opts, params) {
    await plugin_commands_installation_1.install.handler(opts);
    await plugin_commands_script_runners_1.test.handler(opts, params); // tslint:disable-line:no-any
}
exports.handler = handler;
