// @pnpm/plugin-commands-script-runners/lib/start.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cli_utils_1 = require("@pnpm/cli-utils");
const config_1 = require("@pnpm/config");
const common_tags_1 = require("common-tags");
const R = require("ramda");
const renderHelp = require("render-help");
const run_1 = require("./run");
function rcOptionsTypes() {
    return {
        ...R.pick([
            'npm-path',
        ], config_1.types),
    };
}
exports.rcOptionsTypes = rcOptionsTypes;
function cliOptionsTypes() {
    return run_1.IF_PRESENT_OPTION;
}
exports.cliOptionsTypes = cliOptionsTypes;
exports.commandNames = ['start'];
function help() {
    return renderHelp({
        description: common_tags_1.oneLine `
      Runs an arbitrary command specified in the package's "start" property of its "scripts" object.
      If no "start" property is specified on the "scripts" object, it will run node server.js.`,
        descriptionLists: [
            {
                title: 'Options',
                list: [
                    run_1.IF_PRESENT_OPTION_HELP,
                ],
            },
        ],
        url: cli_utils_1.docsUrl('start'),
        usages: ['pnpm start [-- <args>...]'],
    });
}
exports.help = help;
async function handler(opts, params) {
    return run_1.handler(opts, ['start', ...params]);
}
exports.handler = handler;
