// @pnpm/plugin-commands-script-runners/lib/stop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cli_utils_1 = require("@pnpm/cli-utils");
const config_1 = require("@pnpm/config");
const R = require("ramda");
const renderHelp = require("render-help");
const run_1 = require("./run");
function rcOptionsTypes() {
    return {
        ...R.pick([
            'npm-path',
        ], config_1.types),
    };
}
exports.rcOptionsTypes = rcOptionsTypes;
function cliOptionsTypes() {
    return run_1.IF_PRESENT_OPTION;
}
exports.cliOptionsTypes = cliOptionsTypes;
exports.commandNames = ['stop'];
function help() {
    return renderHelp({
        description: `Runs a package's "stop" script, if one was provided.`,
        descriptionLists: [
            {
                title: 'Options',
                list: [
                    run_1.IF_PRESENT_OPTION_HELP,
                ],
            },
        ],
        url: cli_utils_1.docsUrl('stop'),
        usages: ['pnpm stop [-- <args>...]'],
    });
}
exports.help = help;
async function handler(opts, params) {
    return run_1.handler(opts, ['stop', ...params]);
}
exports.handler = handler;
